designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala

```diff
@@ -0,0 +1,160 @@
+        if (paramsMap.keys.size != customActionParams.size) {
+          Left(
+            MismatchedParamsError(
+              s"Validation requires different count of custom action parameters than provided in request for: ${request.actionName}"
+            )
+          )
+        } else { Right(paramsMap) }
+      case (true, None) =>
+        Left(MismatchedParamsError(s"Missing required params for action: ${request.actionName}"))
+      case (false, Some(_)) =>
+        Left(MismatchedParamsError(s"Params found for no params action: ${request.actionName}"))
+      case _ => Right(Map.empty[String, String])
+    }
+  }
+
+  private def validateParams(
+      requestParamsMap: Either[CustomActionValidationError, Map[String, String]],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): Either[CustomActionValidationError, Map[String, List[PartSubGraphCompilationError]]] = {
+    val checkedParamsMap = checkForMissingKeys(requestParamsMap, customActionParams)
+
+    checkedParamsMap.map { m =>
+      m.map { case (k, v) =>
+        (
+          k,
+          customActionParams.find(_.name == k) match {
+            case Some(param) =>
+              param.validators
+                .getOrElse(Nil)
+                .map {
+                  _.isValid(
+                    paramName = ParameterName(k),
+                    expression = Expression.spel("None"),
+                    value = Some(v),
+                    label = None
+                  )
+                }
+                .collect { case Invalid(i) => i }
+            case None =>
+              throw new IllegalStateException
+          }
+        )
+      }
+    }
+  }
+
+  private def checkForMissingKeys(
+      requestParamsMap: Either[CustomActionValidationError, Map[String, String]],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): Either[CustomActionValidationError, Map[String, String]] = {
+    requestParamsMap.flatMap { map =>
+      val nameList    = customActionParams.map(_.name)
+      val missingKeys = nameList.filterNot(map.contains)
+      missingKeys match {
+        case Nil => Right(map)
+        case _ => Left(MismatchedParamsError(s"Missing params: ${missingKeys.mkString(", ")} for action: ${nodeId.id}"))
+      }
+    }
+  }
+
+  private def getValidationResult(
+      validatedParams: Either[CustomActionValidationError, Map[String, List[PartSubGraphCompilationError]]]
+  ): Either[CustomActionValidationError, CustomActionValidationResult] = {
+    val hasErrors = validatedParams.map { m => m.exists { case (_, errorList) => errorList.nonEmpty } }
+
+    hasErrors match {
+      case Right(true) =>
+        Right(CustomActionValidationResult.Invalid(validatedParams.getOrElse(throw new IllegalStateException)))
+      case Right(false) => Right(CustomActionValidationResult.Valid)
+      case Left(va)     => Left(va)
+    }
+  }
+
+  def validateCustomActionParams(
```

Comment:
`validateCustomActionParams` is not in the responsibliity scope of validator, should be moved to DeploymentServiceImpl.

Reply:
moved

Reply:
I still see here `validateCustomActionParams`and `fromCommand`. Lets keep that validator responsible only for validation on action level, and move that translation from command outside, to `DeploymentServiceImpl`.